# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [2]:
import Toolbox.apriorimining


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `apriorimining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [12]:
Toolbox.apriorimining.generate_association_rules()

Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data\courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.33
set= { 6, 7 },  sup= 83.33
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
--------------------------------------------------------


There are 3 association rules: {} -> {6}; {2} -> {6}; {7} -> {6}. The first is ignored because it contains the empty set {}; for the other two it's noticeable that {6} is always on the right. We can see from the graph that every student took Physics, so this isn't surprising. Courses 2 and 7 are Maths and Chemistry, both of which are very related to physics. So these association rules make sense; their confidence is misleading though because everyone took physics in this group.


  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? The script can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [3]:
Toolbox.apriorimining.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data\MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1

The three rules with highest confidence are {Empire Strikes Back, the (ESB), Raiders of the Lost Ark, Return of the Jedi (RotJ)} -> {Star Wars (SW)}; {ESB, RotJ} -> {SW}; and {Pulp Fiction, RotJ} -> {SW}.
These make sense for the most part because ESB and RotJ are sequels to the original SW: so if you watch those then you most likely also watched the original SW. Raiders of the Lost Ark and Pulp Fiction are also kind of related because they share actors/themes with the Star Wars franchise (Raiders more than Pulp Fiction).

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

Star Wars, Contact, and Fargo have been watched the most times. Rules with more than three items have a very low chance of having a high support/confidence so they will be a lot rarer than rules with less items.


#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

Suppose the antecedent has a low frequency: this means the support is relatively low. But if the antecedent always occurs with the item after the arrow, then the confidence will be 100%. Therefore it's very possible to have very low support but still high confidence.

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

1. To find the confidence we divide the support for itemset {a,b} by the support for item a: 30 / 45 = 66.7%. 66.7% > 60% so this rule has high enough confidence to pass the threshold: it's an interesting rule
2. Lift is equal to confidence/P(b): 66.7/80 = 0.84: this is smaller than 1 so there's a negative effect.
3. Even though the confidence is acceptable (66.7%), the nature of the rule is reversed: there's a negative correlation instead of a positive one.

#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

1. c2 <= c1 <= c3 or c2 <= c3 <= c1
c2 is definitely the lowest because now a not only has to occur together with b, but also with c. So the consequent itemset is made larger and the percentage of occurrences of a in comparison to b/c is larger than for c1/c3. As for c1 and c3, it really depends on c which one's confidence is higher.
2. yes, suppose a = physics, b = biology, c = history from 5.1: confidence from biology to history is 0.33 and confidence from physics to biology is 0.5, but the confidence from physics to history is 0.17: if we put a threshold of 0.30 then a -> b and b -> c are above the threshold, while a -> c is below.


#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



1. (105/192) / (40/102) = 1.39 for table 2; (2/11) / (5/25) = 0.91 for college students, (103/181) / (35/77) = 1.25 for working adults. It's a positive correlation for the combined group and adult group, but negative for the college students. This can occur because the group of college students is smaller.
2. phi = 105+62-(40+87) / sqrt(294) = 2.33; 2+20-(5+9)/sqrt(36) = 1.33 for college students, 103+42-(35+78)/sqrt(258) = 1.99 for working adults. While there's still a big difference between the college students group and working adult/mixed group, all values are larger than 1 so the direction doesn't change.
3. lift = (105/192)/(145/294) = 1.11; lift = (2/11)/(7/36) = 0.93 for college students, (103/181)/(138/258) = 1.06 for working adults. Same story as for the odds ratio: mixed group/adult group has positive correlation while it's negative for the college students.